### PR TITLE
[REVIEW] Various optimization across all functions

### PR DESCRIPTION
**Pull request by mnicely**
The PR:
1. Implement a style standard of add C++11 to all Elementwise kernels.
2. Replace `cupy.fftpack` with `cupy.fft`
3. Series of elementwise kernels optimizations.
4. Clean up logic to call CUDA kernels.
5. Move firwin to CPU, expected data size too small (~2x resample_poly)

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by awthomp** [comment]
Hey @mnicely, can you change the title of this PR to account for the fftpack changes too?

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,7 @@
 
 
 ## Improvements
+- PR #267 Add C++11 to Elementwise Kernels
```
Comment (awthomp) [inline CHANGELOG.md:8]: Can you expand this comment that you also replaced cupy.fftpack with cupy.fft

Comment (mnicely) [reply CHANGELOG.md]: Fixed

File: python/cusignal/filter_design/fir_filter_design.py [hunk 2]
```diff
@@ -255,22 +256,22 @@ def firwin(
...
     # is even, and each pair in cutoff corresponds to passband.
-    cutoff = cp.hstack(([0.0] * pass_zero, cutoff, [1.0] * pass_nyquist))
+    cutoff = np.hstack(([0.0] * pass_zero, cutoff, [1.0] * pass_nyquist))
 
     # `bands` is a 2D array; each row gives the left and right edges of
     # a passband.
     bands = cutoff.reshape(-1, 2)
 
     # Build up the coefficients.
     alpha = 0.5 * (numtaps - 1)
```
Comment (awthomp) [inline python/cusignal/filter_design/fir_filter_design.py:266]: This seemingly replaces filter creation from cupy to numpy, but have we benchmarked E2E performance (e.g. applying that filter to in GPU data). I'm curious if the extra H->D copy negates performance you get there?

File: CHANGELOG.md [hunk 3]
```diff
@@ -5,6 +5,7 @@
 
 
 ## Improvements
+- PR #267 Various optimization across all functions
```
Comment (awthomp) [inline CHANGELOG.md:8]: LOL. This is the "stuff with things" comment.

**Review by awthomp — COMMENTED** [review]
Looks good. Only concern is with the fir_filter_design being moved to CPU (numpy) when you look at E2E performance.

**Comment by mnicely** [comment]
> Looks good. Only concern is with the fir_filter_design being moved to CPU (numpy) when you look at E2E performance.

The only transfer that I see is here. Everything else is already on the device.
```python
    # Get and apply the window function.
    win = cp.asnumpy(get_window(window, numtaps, fftbins=False))
    h *= win
```
For some scenarios calling `scipy.signal.get_window` will even be faster. I left in cusignal's just in case the number of taps is large, where compute would be greater than transfer.

I'll test E2E with this PR and branch-0.16 and post results here.

**Review by awthomp — APPROVED** [review]

**Comment by mnicely** [comment]
@awthomp I'm not seeing a negative impact.

### This PR
```bash
%%time
resample_cpu_ensemble = signal.resample_poly(cpu_ensemble, up, down, axis=1, window='flattop')
CPU times: user 2.38 s, sys: 312 ms, total: 2.69 s
Wall time: 2.69 s

%%time
resample_gpu_ensemble = cusignal.resample_poly(gpu_ensemble, up, down, axis=1, window='flattop')
CPU times: user 76.3 ms, sys: 4.02 ms, total: 80.3 ms
Wall time: 79.9 ms
```

### Branch-0.16
```bash
%%time
resample_cpu_ensemble = signal.resample_poly(cpu_ensemble, up, down, axis=1, window='flattop')
CPU times: user 2.38 s, sys: 312 ms, total: 2.69 s
Wall time: 2.69 s

%%time
resample_gpu_ensemble = cusignal.resample_poly(gpu_ensemble, up, down, axis=1, window='flattop')
CPU times: user 82 ms, sys: 4.14 ms, total: 86.1 ms
Wall time: 85.5 ms
```

**Comment by awthomp** [comment]
Thanks for the sanity check, @mnicely. LGTM!

**Pull request merged**